src/fmgc/src/flightphase/FlightPhaseManager.ts

```diff
@@ -0,0 +1,131 @@
+import { Phase, PreFlightPhase, TakeOffPhase, ClimbPhase, CruisePhase, DescentPhase, ApproachPhase, GoAroundPhase, DonePhase } from '@fmgc/flightphase/Phase';
+import { VerticalMode } from '@shared/autopilot';
+import { FmgcFlightPhase } from '@shared/flightphase';
+
+function canInitiateDes(distanceToDestination: number): boolean {
+    const fl = Math.round(Simplane.getAltitude() / 100);
+    const fcuSelFl = Simplane.getAutoPilotDisplayedAltitudeLockValue('feet') / 100;
+    const cruiseFl = SimVar.GetSimVarValue('L:AIRLINER_CRUISE_ALTITUDE', 'number') / 100;
+
+    // Can initiate descent? OR Can initiate early descent?
+    return ((distanceToDestination < 200 || fl < 200) && fcuSelFl < cruiseFl && fcuSelFl < fl)
+        || (distanceToDestination >= 200 && fl > 200 && fcuSelFl <= 200);
+}
+
+export class FlightPhaseManager {
+    private activePhase = SimVar.GetSimVarValue('L:A32NX_INITIAL_FLIGHT_PHASE', 'number') || FmgcFlightPhase.PREFLIGHT;
+
+    private phases: { [key in FmgcFlightPhase]: Phase } = {
+        [FmgcFlightPhase.PREFLIGHT]: new PreFlightPhase(),
+        [FmgcFlightPhase.TAKEOFF]: new TakeOffPhase(),
+        [FmgcFlightPhase.CLIMB]: new ClimbPhase(),
+        [FmgcFlightPhase.CRUISE]: new CruisePhase(),
+        [FmgcFlightPhase.DESCENT]: new DescentPhase(),
+        [FmgcFlightPhase.APPROACH]: new ApproachPhase(),
+        [FmgcFlightPhase.GOAROUND]: new GoAroundPhase(),
+        [FmgcFlightPhase.DONE]: new DonePhase(),
+    }
+
+    private phaseChangeListeners: [(prev: FmgcFlightPhase, next: FmgcFlightPhase) => void];
+
+    get phase() {
+        return this.activePhase;
+    }
+
+    init(): void {
+        console.log(`FMGC Flight Phase: ${this.phase}`);
+        this.phases[this.phase].init();
+    }
+
+    checkPhase(_deltaTime: number): void {
+        if (this.phases[this.phase].check(_deltaTime)) {
+            this.changePhase(this.phases[this.phase].nextPhase);
+        }
+    }
+
+    addOnPhaseChanged(cb: (prev: FmgcFlightPhase, next: FmgcFlightPhase) => void): void {
+        this.phaseChangeListeners.push(cb);
```

Comment:
**question:** is `addOnPhaseChanged` actually called anywhere ?

Reply:
https://github.com/flybywiresim/a32nx/pull/6654/files#diff-1e066709d964d45a102d2029d3b294cbffc322231c19cfb38ff109e973155ad1R257